File: apps/engine/src/agentcy/forecast/brief_v1.py
"""Thin canonical brief.v1 import adapter for agentcy-echo."""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Literal

from pydantic import BaseModel, Field, ValidationError, field_validator, model_validator

ID_PATTERN = re.compile(r"^[a-z0-9][a-z0-9._-]{2,127}$")


class BriefWriter(BaseModel):
    repo: Literal["brand-os"] = "brand-os"
    module: Literal["agentcy-compass"] = "agentcy-compass"


class BriefSignal(BaseModel):
    source: str = Field(min_length=1)
    summary: str = Field(min_length=1)
    evidence: list[str] = Field(default_factory=list)


class BriefStrategy(BaseModel):
    angle: str = Field(min_length=1)
    cta: str = Field(min_length=1)
    platforms: list[str] = Field(min_length=1)


class BriefPolicy(BaseModel):
    verdict: Literal["approved", "escalate", "deny"]
    confidence: float = Field(ge=0, le=1)
    notes: list[str] = Field(default_factory=list)


class BriefDeliverable(BaseModel):
    kind: str = Field(min_length=1)
    channel: str = Field(min_length=1)
    notes: str | None = None


class BriefCreative(BaseModel):
    headline: str = Field(min_length=1)
    copy_text: str = Field(alias="copy", min_length=1)
    tone_notes: list[str] = Field(min_length=1)
    deliverables: list[BriefDeliverable] = Field(default_factory=list)


class BriefLineage(BaseModel):
    source_voice_pack_id: str | None = None
    campaign_id: str | None = None
    signal_id: str | None = None

    @field_validator("source_voice_pack_id", "campaign_id", "signal_id")
    @classmethod
    def _validate_optional_ids(cls, value: str | None) -> str | None:
        if value is None:
            return value
        if not ID_PATTERN.match(value):
            raise ValueError("lineage IDs must match canonical pattern")
        return value


class BriefV1(BaseModel):
    artifact_type: Literal["brief.v1"]
    schema_version: Literal["v1"]
    brief_id: str
    brand_id: str
    voice_pack_id: str
    writer: BriefWriter
    objective: str = Field(min_length=1)
    signal: BriefSignal
    strategy: BriefStrategy
    policy: BriefPolicy
    creative: BriefCreative
    lineage: BriefLineage | None = None

    @field_validator("brief_id", "brand_id", "voice_pack_id")
    @classmethod
    def _validate_ids(cls, value: str) -> str:
        if not ID_PATTERN.match(value):
            raise ValueError("IDs must match canonical pattern")
        return value

    @model_validator(mode="after")
    def _validate_lineage(self) -> BriefV1:
        if self.lineage and self.lineage.source_voice_pack_id is not None:
            if self.lineage.source_voice_pack_id != self.voice_pack_id:
                raise ValueError("lineage.source_voice_pack_id must equal voice_pack_id")
        return self


class ImportedBrief(BaseModel):
    brief: BriefV1
    requirement: str
    frozen_input: dict
    lineage: dict



def _clean_lines(values: list[str]) -> list[str]:
    cleaned: list[str] = []
    seen: set[str] = set()
    for value in values:
        normalized = value.strip()
        if not normalized or normalized in seen:
            continue
        seen.add(normalized)
        cleaned.append(normalized)
    return cleaned



def _derive_requirement(brief: BriefV1) -> str:
    platforms = ", ".join(brief.strategy.platforms)
    deliverable_channels = ", ".join(
        _clean_lines([item.channel for item in brief.creative.deliverables])
    )
    parts = _clean_lines(
        [
            f"Objective: {brief.objective}",
            f"Signal source: {brief.signal.source}",
            f"Signal summary: {brief.signal.summary}",
            f"Strategy angle: {brief.strategy.angle}",
            f"Call to action: {brief.strategy.cta}",
            f"Primary platforms: {platforms}" if platforms else "",
            f"Creative headline: {brief.creative.headline}",
            f"Creative copy seed: {brief.creative.copy_text}",
            f"Deliverable channels: {deliverable_channels}" if deliverable_channels else "",
            "Model plausible public reaction, likely discussion themes, and platform-specific response dynamics.",
        ]
    )
    return "\n".join(parts)



def _build_frozen_input(brief: BriefV1) -> dict:
    return {
        "artifact_type": brief.artifact_type,
        "schema_version": brief.schema_version,
        "brief_id": brief.brief_id,
        "brand_id": brief.brand_id,
        "voice_pack_id": brief.voice_pack_id,
        "objective": brief.objective,
        "signal": brief.signal.model_dump(exclude_none=True),
        "strategy": brief.strategy.model_dump(exclude_none=True),
        "policy": brief.policy.model_dump(exclude_none=True),
        "creative": brief.creative.model_dump(exclude_none=True),
        "lineage": {
            "source_brief_id": brief.brief_id,
            "source_voice_pack_id": (
                brief.lineage.source_voice_pack_id if brief.lineage else brief.voice_pack_id
            ),
            "campaign_id": brief.lineage.campaign_id if brief.lineage else None,
            "signal_id": brief.lineage.signal_id if brief.lineage else None,
        },
    }



def _build_manifest_lineage(brief: BriefV1) -> dict:
    return {
        "brief_id": brief.brief_id,
        "brand_id": brief.brand_id,
        "source_voice_pack_id": (
            brief.lineage.source_voice_pack_id if brief.lineage else brief.voice_pack_id
        ),
        "campaign_id": brief.lineage.campaign_id if brief.lineage else None,
        "signal_id": brief.lineage.signal_id if brief.lineage else None,
    }



def import_brief_v1(path: str | Path) -> ImportedBrief:
    brief_path = Path(path).expanduser().resolve()
    try:
        payload = json.loads(brief_path.read_text(encoding="utf-8"))
    except FileNotFoundError:
        raise
    except json.JSONDecodeError as exc:
        raise ValueError(f"Invalid brief.v1 JSON: {exc}") from exc

    try:
        brief = BriefV1.model_validate(payload)
    except ValidationError as exc:
        raise ValueError(f"Invalid brief.v1 payload: {exc}") from exc

    return ImportedBrief(
        brief=brief,
        requirement=_derive_requirement(brief),
        frozen_input=_build_frozen_input(brief),
        lineage=_build_manifest_lineage(brief),
    )

File: apps/engine/src/studio/runtime/src/domain/brief-v1.ts
import { readFileSync } from 'fs'

export interface CanonicalBriefV1 {
  artifact_type: 'brief.v1'
  schema_version: 'v1'
  brief_id: string
  brand_id: string
  voice_pack_id: string
  writer: {
    repo: 'brand-os'
    module: 'agentcy-compass'
  }
  objective: string
  signal: {
    source: string
    summary: string
    evidence?: string[]
  }
  strategy: {
    angle: string
    cta: string
    platforms: string[]
  }
  policy: {
    verdict: 'approved' | 'escalate' | 'deny'
    confidence: number
    notes?: string[]
  }
  creative: {
    headline: string
    copy: string
    tone_notes: string[]
    deliverables?: Array<{
      kind: string
      channel: string
      notes?: string
    }>
  }
  lineage?: {
    source_voice_pack_id?: string
    campaign_id?: string
    signal_id?: string
  }
}

function expectRecord(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error(`Invalid brief.v1: ${label} must be an object`)
  }
  return value as Record<string, unknown>
}

function expectNonEmptyString(value: unknown, label: string): string {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw new Error(`Invalid brief.v1: ${label} must be a non-empty string`)
  }
  return value.trim()
}

function expectStringArray(value: unknown, label: string, options: { minItems?: number } = {}): string[] {
  if (!Array.isArray(value)) {
    throw new Error(`Invalid brief.v1: ${label} must be an array`)
  }
  const strings = value.map((item, index) => expectNonEmptyString(item, `${label}[${index}]`))
  if ((options.minItems ?? 0) > strings.length) {
    throw new Error(`Invalid brief.v1: ${label} must contain at least ${options.minItems} item(s)`)
  }
  return strings
}

export function parseBriefV1(input: unknown): CanonicalBriefV1 {
  const brief = expectRecord(input, 'brief.v1 payload')
  if (brief.artifact_type !== 'brief.v1') {
    throw new Error('Invalid brief.v1: artifact_type must equal "brief.v1"')
  }
  if (brief.schema_version !== 'v1') {
    throw new Error('Invalid brief.v1: schema_version must equal "v1"')
  }

  const writer = expectRecord(brief.writer, 'writer')
  if (writer.repo !== 'brand-os' || writer.module !== 'agentcy-compass') {
    throw new Error('Invalid brief.v1: writer must be brand-os / agentcy-compass')
  }

  const signal = expectRecord(brief.signal, 'signal')
  const strategy = expectRecord(brief.strategy, 'strategy')
  const policy = expectRecord(brief.policy, 'policy')
  const creative = expectRecord(brief.creative, 'creative')
  const lineage = brief.lineage === undefined ? undefined : expectRecord(brief.lineage, 'lineage')

  const parsed: CanonicalBriefV1 = {
    artifact_type: 'brief.v1',
    schema_version: 'v1',
    brief_id: expectNonEmptyString(brief.brief_id, 'brief_id'),
    brand_id: expectNonEmptyString(brief.brand_id, 'brand_id'),
    voice_pack_id: expectNonEmptyString(brief.voice_pack_id, 'voice_pack_id'),
    writer: {
      repo: 'brand-os',
      module: 'agentcy-compass',
    },
    objective: expectNonEmptyString(brief.objective, 'objective'),
    signal: {
      source: expectNonEmptyString(signal.source, 'signal.source'),
      summary: expectNonEmptyString(signal.summary, 'signal.summary'),
      evidence: signal.evidence === undefined ? undefined : expectStringArray(signal.evidence, 'signal.evidence'),
    },
    strategy: {
      angle: expectNonEmptyString(strategy.angle, 'strategy.angle'),
      cta: expectNonEmptyString(strategy.cta, 'strategy.cta'),
      platforms: expectStringArray(strategy.platforms, 'strategy.platforms', { minItems: 1 }),
    },
    policy: {
      verdict: (() => {
        const verdict = expectNonEmptyString(policy.verdict, 'policy.verdict')
        if (!['approved', 'escalate', 'deny'].includes(verdict)) {
          throw new Error('Invalid brief.v1: policy.verdict must be one of approved, escalate, deny')
        }
        return verdict as CanonicalBriefV1['policy']['verdict']
      })(),
      confidence: (() => {
        if (typeof policy.confidence !== 'number' || Number.isNaN(policy.confidence) || policy.confidence < 0 || policy.confidence > 1) {
          throw new Error('Invalid brief.v1: policy.confidence must be a number between 0 and 1')
        }
        return policy.confidence
      })(),
      notes: policy.notes === undefined ? undefined : expectStringArray(policy.notes, 'policy.notes'),
    },
    creative: {
      headline: expectNonEmptyString(creative.headline, 'creative.headline'),
      copy: expectNonEmptyString(creative.copy, 'creative.copy'),
      tone_notes: expectStringArray(creative.tone_notes, 'creative.tone_notes', { minItems: 1 }),
      deliverables: undefined,
    },
    lineage: lineage === undefined
      ? undefined
      : {
        source_voice_pack_id: lineage.source_voice_pack_id === undefined
          ? undefined
          : expectNonEmptyString(lineage.source_voice_pack_id, 'lineage.source_voice_pack_id'),
        campaign_id: lineage.campaign_id === undefined
          ? undefined
          : expectNonEmptyString(lineage.campaign_id, 'lineage.campaign_id'),
        signal_id: lineage.signal_id === undefined
          ? undefined
          : expectNonEmptyString(lineage.signal_id, 'lineage.signal_id'),
      },
  }

  if (creative.deliverables !== undefined) {
    if (!Array.isArray(creative.deliverables)) {
      throw new Error('Invalid brief.v1: creative.deliverables must be an array')
    }
    parsed.creative.deliverables = creative.deliverables.map((item, index) => {
      const deliverable = expectRecord(item, `creative.deliverables[${index}]`)
      return {
        kind: expectNonEmptyString(deliverable.kind, `creative.deliverables[${index}].kind`),
        channel: expectNonEmptyString(deliverable.channel, `creative.deliverables[${index}].channel`),
        notes: deliverable.notes === undefined ? undefined : expectNonEmptyString(deliverable.notes, `creative.deliverables[${index}].notes`),
      }
    })
  }

  if (parsed.lineage && parsed.lineage.source_voice_pack_id === undefined) {
    throw new Error('Invalid brief.v1: lineage.source_voice_pack_id is required when lineage is present')
  }
  if (parsed.lineage?.source_voice_pack_id && parsed.lineage.source_voice_pack_id !== parsed.voice_pack_id) {
    throw new Error('Invalid brief.v1: lineage.source_voice_pack_id must equal voice_pack_id')
  }

  return parsed
}

export function loadBriefV1(path: string): CanonicalBriefV1 {
  let payload: unknown
  try {
    payload = JSON.parse(readFileSync(path, 'utf8')) as unknown
  } catch (error) {
    throw new Error(`Invalid brief.v1: failed to parse JSON at ${path}: ${error instanceof Error ? error.message : String(error)}`)
  }
  return parseBriefV1(payload)
}
